### DISCUSSION TITLE
Sidekiq.dump_json vs YAML.dump

### DISCUSSION THREAD

**Discussion by mcieaaka** [sidekiq/sidekiq#6340] [Q&A]
Hi, I recently implemented `Sidekiq::Batch` in my application using Sidekiq Pro. 
The scenario was as such: I had a long running job which updates records in bulk and at the end there was a Delayed Notification Job. So I implemented Batch to improve performance of the Bulk Update job and moved the Notification Job to a callback.

While implementing, I faced a very peculiar issue. The arguments being passed to the `On Success` callback did not work as expected. I was passing an `ActiveSupport::TimeWithZone` object as one of the arguments in my callback. But in the callback the "Time" argument lost its precision as well as got converted to string. After experimenting and looking at Sidekiq's internal working of workers and callbacks I found the difference:
In Sidekiq::Batch we are using `Sidekiq.dump_json` to parse arguments whereas in Worker we are using `YAML.dump` and this created all the difference.
In order to fix this, I simply changed my working of Notification Job because I didn't really need the TimeWithZone object. But I was curious why two different mechanisms were used to parse arguments.
Note: I am using Sidekiq v6.4

```
Loading test environment (Rails 6.1.7)
[1] pry(main)> args = {a: "some_text", b: Time.zone.now}
=> {:a=>"some_text", :b=>Fri, 21 Jun 2024 11:55:18.161535284 IST +05:30}
[2] pry(main)> args = Sidekiq.dump_json(args)
=> "{\"a\":\"some_text\",\"b\":\"2024-06-21 11:55:18 +0530\"}"
[3] pry(main)> args = Sidekiq.load_json(args)
=> {"a"=>"some_text", "b"=>"2024-06-21 11:55:18 +0530"}
[4] pry(main)> args2 = {a: "some_text", b: Time.zone.now}
=> {:a=>"some_text", :b=>Fri, 21 Jun 2024 11:56:31.822981464 IST +05:30}
[5] pry(main)> args2 = YAML.dump(args2)
=> "---\n:a: some_text\n:b: !ruby/object:ActiveSupport::TimeWithZone\n  utc: 2024-06-21 06:26:31.822981464 Z\n  zone: !ruby/object:ActiveSupport::TimeZone\n    name: Asia/Kolkata\n  time: 2024-06-21 11:56:31.822981464 Z\n"
[6] pry(main)> args2 = YAML.load(args2)
=> {:a=>"some_text", :b=>Fri, 21 Jun 2024 11:56:31.822981464 IST +05:30}
[7] pry(main)>
```
![image](https://github.com/sidekiq/sidekiq/assets/58990472/1a04b2e5-27bf-4aa8-bf32-f66bdc6eac7d)

**Comment by mperham** [comment]
Since forever, Sidekiq only supports JSON-safe types for job and callback arguments. I don't believe there's anywhere in Sidekiq where we use YAML.dump/load.

**Reply by mcieaaka** [reply]
Thanks for replying @mperham, 
Yup you are right, for callbacks and workers we do use JSON safe args. However, YAML is used in the Delayed extensions, which was what I used in my app as well. I was using "delay" for my notification job as mentioned in the main text.
I also just realised that the use of YAML is probably the reason why Sidekiq has discontinued delayed extensions. 
I referred the following wiki: https://github.com/sidekiq/sidekiq/wiki/Delayed-extensions
Here's the path for reference: BRANCH: 6-x File path: lib/sidekiq/extensions/action_mailer.rb
![image](https://github.com/sidekiq/sidekiq/assets/58990472/4c32dd25-a125-4fb3-9b19-c02e85b0e069)
Anyways I got my answer, Thanks!
